File: src/openagent/core/context_budget.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .message_materializer import materialize_payload
from .token_counting import count_materialized_payload
from .types import ChatMessage, Model, ToolSchema

DEFAULT_BYTES_PER_TOKEN = 3
DEFAULT_GUARD_RATIO = 0.9
DEFAULT_INPUT_SAFETY_MARGIN_TOKENS = 1024
DEFAULT_TOOL_DISPLAY_MAX_BYTES = 50 * 1024
DEFAULT_TOOL_CONTEXT_PREVIEW_BYTES = 4096
DEFAULT_TOOL_CONTEXT_PREVIEW_LINES = 40
DEFAULT_TOOL_CONTEXT_LINE_MAX_CHARS = 240
DEFAULT_PRUNE_OLD_TOOL_OUTPUTS = True
DEFAULT_PRUNE_KEEP_RECENT_USER_TURNS = 2
DEFAULT_PRUNE_PROTECT_INPUT_TOKENS = 12_000
DEFAULT_PRUNE_MIN_INPUT_TOKENS = 4_000
DEFAULT_COMPACT_SUMMARY_MAX_OUTPUT_TOKENS = 512
DEFAULT_COMPACT_REFRESH_MIN_NEW_MESSAGES = 6
DEFAULT_OVERFLOW_KEEP_RECENT_USER_TURNS = 2
SUPPORTED_STRATEGIES = {"auto", "error", "compact"}
SUPPORTED_COUNTING = {"auto", "tiktoken", "heuristic"}


@dataclass(frozen=True, slots=True)
class ContextBudgetResult:
    estimated_input_tokens: int
    input_limit_tokens: int
    context_window: int
    reserved_output_tokens: int
    overflowed: bool
    tool_message_count: int = 0
    largest_tool_message_tokens: int = 0
    largest_tool_message_name: str = ""
    counting_method: str = "heuristic"
    counting_exact: bool = False
    fallback_stage: str = "initial"
    payload_kind: str = "generic"


class ContextBudgetConfigError(ValueError):
    """Raised when context budget options are invalid or unsupported."""


def check_context_budget(
    *,
    system: str | None,
    messages: list[ChatMessage],
    tools: list[ToolSchema],
    model: Model | None,
    options: dict[str, Any] | None = None,
    fallback_stage: str = "initial",
) -> ContextBudgetResult | None:
    if model is None or model.context_window <= 0:
        return None

    config = load_context_budget_options(options, model=model)
    if not config["enabled"]:
        return None

    strategy = config["strategy"]
    if strategy not in SUPPORTED_STRATEGIES:
        raise ContextBudgetConfigError(
            f"Unsupported context budget strategy: {strategy}. "
            "Supported strategies: auto, error, compact."
        )

    counting = config["counting"]
    if counting not in SUPPORTED_COUNTING:
        raise ContextBudgetConfigError(
            f"Unsupported context budget counting mode: {counting}. "
            "Supported modes: auto, tiktoken, heuristic."
        )

    reserved_output_tokens = int(config["reserve_output_tokens"])
    input_limit_tokens = _compute_input_limit_tokens(model=model, config=config)
    materialized = materialize_payload(
        system=system,
        messages=messages,
        tools=tools,
        model=model,
        options=options,
    )
    count = count_materialized_payload(
        materialized,
        model=model,
        options=options,
        counting=counting,
        bytes_per_token=int(config["bytes_per_token"]),
    )
    diagnostics = _tool_message_diagnostics(
        messages=messages,
        bytes_per_token=int(config["bytes_per_token"]),
        model=model,
        options=options,
        counting=counting,
    )
    return ContextBudgetResult(
        estimated_input_tokens=count.tokens,
        input_limit_tokens=input_limit_tokens,
        context_window=model.context_window,
        reserved_output_tokens=reserved_output_tokens,
        overflowed=count.tokens > input_limit_tokens,
        tool_message_count=diagnostics["tool_message_count"],
        largest_tool_message_tokens=diagnostics["largest_tool_message_tokens"],
        largest_tool_message_name=diagnostics["largest_tool_message_name"],
        counting_method=count.method,
        counting_exact=count.exact,
        fallback_stage=fallback_stage,
        payload_kind=materialized.payload_kind,
    )


def format_context_budget_error(result: ContextBudgetResult) -> str:
    message = (
        "Context budget exceeded before model call: "
        f"estimated_input_tokens={result.estimated_input_tokens}, "
        f"input_limit_tokens={result.input_limit_tokens}, "
        f"context_window={result.context_window}, "
        f"reserved_output_tokens={result.reserved_output_tokens}, "
        f"counting_method={result.counting_method}, "
        f"counting_exact={str(result.counting_exact).lower()}, "
        f"payload_kind={result.payload_kind}, "
        f"fallback_stage={result.fallback_stage}"
    )
    if result.tool_message_count > 0:
        message += (
            f", tool_message_count={result.tool_message_count}, "
            f"largest_tool_message_tokens={result.largest_tool_message_tokens}, "
            f"largest_tool_message_name={result.largest_tool_message_name or 'unknown'}"
        )
    return message


def load_context_budget_options(
    options: dict[str, Any] | None,
    *,
    model: Model | None,
) -> dict[str, Any]:
    raw_options = options or {}
    raw_context_budget = raw_options.get("context_budget", {})
    if not isinstance(raw_context_budget, dict):
        raise ContextBudgetConfigError("AgentConfig.options['context_budget'] must be a dict.")

    enabled = _expect_bool(raw_context_budget.get("enabled", True), field_name="enabled")
    strategy = raw_context_budget.get("strategy", "auto")
    if not isinstance(strategy, str) or not strategy.strip():
        raise ContextBudgetConfigError("context_budget.strategy must be a non-empty string.")

    counting = raw_context_budget.get("counting", "auto")
    if not isinstance(counting, str) or not counting.strip():
        raise ContextBudgetConfigError("context_budget.counting must be a non-empty string.")

    reserve_output_tokens = raw_context_budget.get("reserve_output_tokens", model.max_output if model is not None else 0)
    guard_ratio = raw_context_budget.get("guard_ratio", DEFAULT_GUARD_RATIO)
    explicit_input_safety_margin_tokens = "input_safety_margin_tokens" in raw_context_budget
    use_safety_margin_tokens = explicit_input_safety_margin_tokens or "guard_ratio" not in raw_context_budget
    safety_margin_default = DEFAULT_INPUT_SAFETY_MARGIN_TOKENS if use_safety_margin_tokens else 0
    input_safety_margin_tokens = raw_context_budget.get("input_safety_margin_tokens", safety_margin_default)
    bytes_per_token = raw_context_budget.get("bytes_per_token", DEFAULT_BYTES_PER_TOKEN)
    tool_display_max_bytes = raw_context_budget.get("tool_display_max_bytes", DEFAULT_TOOL_DISPLAY_MAX_BYTES)
    tool_context_preview_bytes = raw_context_budget.get("tool_context_preview_bytes", DEFAULT_TOOL_CONTEXT_PREVIEW_BYTES)
    tool_context_preview_lines = raw_context_budget.get("tool_context_preview_lines", DEFAULT_TOOL_CONTEXT_PREVIEW_LINES)
    tool_context_line_max_chars = raw_context_budget.get("tool_context_line_max_chars", DEFAULT_TOOL_CONTEXT_LINE_MAX_CHARS)
    prune_old_tool_outputs = raw_context_budget.get("prune_old_tool_outputs", DEFAULT_PRUNE_OLD_TOOL_OUTPUTS)
    prune_keep_recent_user_turns = raw_context_budget.get("prune_keep_recent_user_turns", DEFAULT_PRUNE_KEEP_RECENT_USER_TURNS)
    prune_protect_input_tokens = raw_context_budget.get("prune_protect_input_tokens", DEFAULT_PRUNE_PROTECT_INPUT_TOKENS)
    prune_min_input_tokens = raw_context_budget.get("prune_min_input_tokens", DEFAULT_PRUNE_MIN_INPUT_TOKENS)
    compact_summary_max_output_tokens = raw_context_budget.get(
        "compact_summary_max_output_tokens",
        DEFAULT_COMPACT_SUMMARY_MAX_OUTPUT_TOKENS,
    )
    compact_refresh_min_new_messages = raw_context_budget.get(
        "compact_refresh_min_new_messages",
        DEFAULT_COMPACT_REFRESH_MIN_NEW_MESSAGES,
    )
    overflow_keep_recent_user_turns = raw_context_budget.get(
        "overflow_keep_recent_user_turns",
        DEFAULT_OVERFLOW_KEEP_RECENT_USER_TURNS,
    )
    overflow_disable_tools_on_final_attempt = raw_context_budget.get("overflow_disable_tools_on_final_attempt", True)
    overflow_final_max_output_tokens = raw_context_budget.get(
        "overflow_final_max_output_tokens",
        min(512, model.max_output) if model is not None else 512,
    )

    reserve_output_tokens = _expect_int(
        reserve_output_tokens,
        field_name="reserve_output_tokens",
        minimum=0,
    )
    guard_ratio = _expect_float(
        guard_ratio,
        field_name="guard_ratio",
        minimum=0.0,
        maximum=1.0,
        include_minimum=False,
    )
    input_safety_margin_tokens = _expect_int(
        input_safety_margin_tokens,
        field_name="input_safety_margin_tokens",
        minimum=0,
    )
    bytes_per_token = _expect_int(bytes_per_token, field_name="bytes_per_token", minimum=1)
    tool_display_max_bytes = _expect_int(tool_display_max_bytes, field_name="tool_display_max_bytes", minimum=1)
    tool_context_preview_bytes = _expect_int(
        tool_context_preview_bytes,
        field_name="tool_context_preview_bytes",
        minimum=1,
    )
    tool_context_preview_lines = _expect_int(
        tool_context_preview_lines,
        field_name="tool_context_preview_lines",
        minimum=1,
    )
    tool_context_line_max_chars = _expect_int(
        tool_context_line_max_chars,
        field_name="tool_context_line_max_chars",
        minimum=1,
    )
    prune_old_tool_outputs = _expect_bool(prune_old_tool_outputs, field_name="prune_old_tool_outputs")
    prune_keep_recent_user_turns = _expect_int(
        prune_keep_recent_user_turns,
        field_name="prune_keep_recent_user_turns",
        minimum=1,
    )
    prune_protect_input_tokens = _expect_int(
        prune_protect_input_tokens,
        field_name="prune_protect_input_tokens",
        minimum=0,
    )
    prune_min_input_tokens = _expect_int(
        prune_min_input_tokens,
        field_name="prune_min_input_tokens",
        minimum=0,
    )
    compact_summary_max_output_tokens = _expect_int(
        compact_summary_max_output_tokens,
        field_name="compact_summary_max_output_tokens",
        minimum=1,
    )
    compact_refresh_min_new_messages = _expect_int(
        compact_refresh_min_new_messages,
        field_name="compact_refresh_min_new_messages",
        minimum=1,
    )
    overflow_keep_recent_user_turns = _expect_int(
        overflow_keep_recent_user_turns,
        field_name="overflow_keep_recent_user_turns",
        minimum=1,
    )
    overflow_disable_tools_on_final_attempt = _expect_bool(
        overflow_disable_tools_on_final_attempt,
        field_name="overflow_disable_tools_on_final_attempt",
    )
    overflow_final_max_output_tokens = _expect_int(
        overflow_final_max_output_tokens,
        field_name="overflow_final_max_output_tokens",
        minimum=1,
    )
    return {
        "enabled": enabled,
        "strategy": strategy.strip(),
        "counting": counting.strip(),
        "reserve_output_tokens": reserve_output_tokens,
        "guard_ratio": guard_ratio,
        "input_safety_margin_tokens": input_safety_margin_tokens,
        "use_safety_margin_tokens": use_safety_margin_tokens,
        "explicit_input_safety_margin_tokens": explicit_input_safety_margin_tokens,
        "bytes_per_token": bytes_per_token,
        "tool_display_max_bytes": tool_display_max_bytes,
        "tool_context_preview_bytes": tool_context_preview_bytes,
        "tool_context_preview_lines": tool_context_preview_lines,
        "tool_context_line_max_chars": tool_context_line_max_chars,
        "prune_old_tool_outputs": prune_old_tool_outputs,
        "prune_keep_recent_user_turns": prune_keep_recent_user_turns,
        "prune_protect_input_tokens": prune_protect_input_tokens,
        "prune_min_input_tokens": prune_min_input_tokens,
        "compact_summary_max_output_tokens": compact_summary_max_output_tokens,
        "compact_refresh_min_new_messages": compact_refresh_min_new_messages,
        "overflow_keep_recent_user_turns": overflow_keep_recent_user_turns,
        "overflow_disable_tools_on_final_attempt": overflow_disable_tools_on_final_attempt,
        "overflow_final_max_output_tokens": overflow_final_max_output_tokens,
    }


def estimate_message_tokens(
    message: ChatMessage,
    *,
    bytes_per_token: int = DEFAULT_BYTES_PER_TOKEN,
    model: Model | None = None,
    options: dict[str, Any] | None = None,
    counting: str = "heuristic",
) -> int:
    materialized = materialize_payload(
        system=None,
        messages=[message],
        tools=[],
        model=model,
        options=options,
    )
    result = count_materialized_payload(
        materialized,
        model=model,
        options=options,
        counting=counting,
        bytes_per_token=bytes_per_token,
    )
    return result.tokens


def _compute_input_limit_tokens(*, model: Model, config: dict[str, Any]) -> int:
    reserved_output_tokens = int(config["reserve_output_tokens"])
    if config["use_safety_margin_tokens"]:
        limit = model.context_window - reserved_output_tokens - int(config["input_safety_margin_tokens"])
        if limit > 0 or bool(config.get("explicit_input_safety_margin_tokens")):
            return max(limit, 0)
    return max(int(model.context_window * float(config["guard_ratio"])) - reserved_output_tokens, 0)


def _tool_message_diagnostics(
    *,
    messages: list[ChatMessage],
    bytes_per_token: int,
    model: Model | None,
    options: dict[str, Any] | None,
    counting: str,
) -> dict[str, Any]:
    tool_message_count = 0
    largest_tool_message_tokens = 0
    largest_tool_message_name = ""
    for message in messages:
        if message.role != "tool":
            continue
        tool_message_count += 1
        estimate = estimate_message_tokens(
            message,
            bytes_per_token=bytes_per_token,
            model=model,
            options=options,
            counting=counting,
        )
        if estimate > largest_tool_message_tokens:
            largest_tool_message_tokens = estimate
            largest_tool_message_name = message.name or ""
    return {
        "tool_message_count": tool_message_count,
        "largest_tool_message_tokens": largest_tool_message_tokens,
        "largest_tool_message_name": largest_tool_message_name,
    }


def _expect_bool(value: Any, *, field_name: str) -> bool:
    if isinstance(value, bool):
        return value
    raise ContextBudgetConfigError(f"context_budget.{field_name} must be a bool.")


def _expect_int(value: Any, *, field_name: str, minimum: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int):
        raise ContextBudgetConfigError(f"context_budget.{field_name} must be an int.")
    if value < minimum:
        raise ContextBudgetConfigError(f"context_budget.{field_name} must be >= {minimum}.")
    return value


def _expect_float(
    value: Any,
    *,
    field_name: str,
    minimum: float,
    maximum: float,
    include_minimum: bool = True,
) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ContextBudgetConfigError(f"context_budget.{field_name} must be a number.")
    number = float(value)
    if include_minimum:
        if number < minimum:
            raise ContextBudgetConfigError(f"context_budget.{field_name} must be >= {minimum}.")
    elif number <= minimum:
        raise ContextBudgetConfigError(f"context_budget.{field_name} must be > {minimum}.")
    if number > maximum:
        raise ContextBudgetConfigError(f"context_budget.{field_name} must be <= {maximum}.")
    return number

File: src/core/context_budget.rs
#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct ContextBudgetOptions {
    pub enabled: bool,
    pub strategy: String,
    pub counting: String,
    pub compaction_mode: String,
    pub reserve_output_tokens: u64,
    pub guard_ratio: f64,
    pub input_safety_margin_tokens: u64,
    pub use_safety_margin_tokens: bool,
    pub explicit_input_safety_margin_tokens: bool,
    pub bytes_per_token: u64,
    pub tool_display_max_bytes: u64,
    pub tool_context_preview_bytes: u64,
    pub tool_context_preview_lines: u64,
    pub tool_context_line_max_chars: u64,
    pub prune_old_tool_outputs: bool,
    pub prune_keep_recent_user_turns: u64,
    pub prune_protect_input_tokens: u64,
    pub prune_min_input_tokens: u64,
    pub compact_summary_max_output_tokens: u64,
    pub compact_refresh_min_new_messages: u64,
    pub overflow_keep_recent_user_turns: u64,
    pub overflow_disable_tools_on_final_attempt: bool,
    pub overflow_final_max_output_tokens: u64,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
pub struct ContextBudgetResult {
    pub estimated_input_tokens: u64,
    pub input_limit_tokens: u64,
    pub context_window: u64,
    pub reserved_output_tokens: u64,
    pub overflowed: bool,
    pub tool_message_count: u64,
    pub largest_tool_message_tokens: u64,
    pub largest_tool_message_name: String,
    pub counting_method: String,
    pub counting_exact: bool,
    pub fallback_stage: String,
    pub payload_kind: String,
}

#[must_use]
pub fn format_context_budget_error(result: &ContextBudgetResult) -> String {
    let mut message = format!(
        "Context budget exceeded before model call: estimated_input_tokens={}, input_limit_tokens={}, context_window={}, reserved_output_tokens={}, counting_method={}, counting_exact={}, payload_kind={}, fallback_stage={}",
        result.estimated_input_tokens,
        result.input_limit_tokens,
        result.context_window,
        result.reserved_output_tokens,
        result.counting_method,
        result.counting_exact,
        result.payload_kind,
        result.fallback_stage
    );
    if result.tool_message_count > 0 {
        message.push_str(&format!(
            ", tool_message_count={}, largest_tool_message_tokens={}, largest_tool_message_name={}",
            result.tool_message_count,
            result.largest_tool_message_tokens,
            if result.largest_tool_message_name.is_empty() {
                "unknown"
            } else {
                &result.largest_tool_message_name
            }
        ));
    }
    message
}

pub fn load_context_budget_options(
    options: Option<&Value>,
    model: Option<&Model>,
) -> Result<ContextBudgetOptions, String> {
    let merged = merge_compaction_facade_options(options)?;
    let enabled = expect_bool(
        merged.get("enabled").unwrap_or(&Value::Bool(true)),
        "enabled",
        "context_budget",
    )?;
    let strategy = expect_non_empty_string(
        merged
            .get("strategy")
            .unwrap_or(&Value::String("auto".to_string())),
        "context_budget.strategy",
    )?;
    let counting = expect_non_empty_string(
        merged
            .get("counting")
            .unwrap_or(&Value::String("auto".to_string())),
        "context_budget.counting",
    )?;
    let compaction_mode = expect_non_empty_string(
        merged
            .get("compaction_mode")
            .unwrap_or(&Value::String(DEFAULT_COMPACTION_MODE.to_string())),
        "context_budget.compaction_mode",
    )?;
    if !SUPPORTED_COMPACTION_MODES.contains(&compaction_mode.as_str()) {
        return Err(format!(
            "Unsupported context_budget.compaction_mode: {compaction_mode}. Supported modes: structured_work_state."
        ));
    }

    let model_max_output = model.map(|item| item.max_output).unwrap_or(0);
    let reserve_output_tokens = expect_int(
        merged
            .get("reserve_output_tokens")
            .unwrap_or(&json!(model_max_output)),
        "reserve_output_tokens",
        0,
        "context_budget",
    )?;
    let guard_ratio = expect_float(
        merged
            .get("guard_ratio")
            .unwrap_or(&json!(DEFAULT_GUARD_RATIO)),
        "guard_ratio",
        0.0,
        1.0,
        false,
    )?;
    let explicit_input_safety_margin_tokens = merged.contains_key("input_safety_margin_tokens");
    let use_safety_margin_tokens =
        explicit_input_safety_margin_tokens || !merged.contains_key("guard_ratio");
    let safety_margin_default = if use_safety_margin_tokens {
        DEFAULT_INPUT_SAFETY_MARGIN_TOKENS
    } else {
        0
    };
    let input_safety_margin_tokens = expect_int(
        merged
            .get("input_safety_margin_tokens")
            .unwrap_or(&json!(safety_margin_default)),
        "input_safety_margin_tokens",
        0,
        "context_budget",
    )?;
    let bytes_per_token = expect_int(
        merged
            .get("bytes_per_token")
            .unwrap_or(&json!(DEFAULT_BYTES_PER_TOKEN)),
        "bytes_per_token",
        1,
        "context_budget",
    )?;
    let tool_display_max_bytes = expect_int(
        merged
            .get("tool_display_max_bytes")
            .unwrap_or(&json!(DEFAULT_TOOL_DISPLAY_MAX_BYTES)),
        "tool_display_max_bytes",
        1,
        "context_budget",
    )?;
    let tool_context_preview_bytes = expect_int(
        merged
            .get("tool_context_preview_bytes")
            .unwrap_or(&json!(DEFAULT_TOOL_CONTEXT_PREVIEW_BYTES)),
        "tool_context_preview_bytes",
        1,
        "context_budget",
    )?;
    let tool_context_preview_lines = expect_int(
        merged
            .get("tool_context_preview_lines")
            .unwrap_or(&json!(DEFAULT_TOOL_CONTEXT_PREVIEW_LINES)),
        "tool_context_preview_lines",
        1,
        "context_budget",
    )?;
    let tool_context_line_max_chars = expect_int(
        merged
            .get("tool_context_line_max_chars")
            .unwrap_or(&json!(DEFAULT_TOOL_CONTEXT_LINE_MAX_CHARS)),
        "tool_context_line_max_chars",
        1,
        "context_budget",
    )?;
    let prune_old_tool_outputs = expect_bool(
        merged
            .get("prune_old_tool_outputs")
            .unwrap_or(&Value::Bool(DEFAULT_PRUNE_OLD_TOOL_OUTPUTS)),
        "prune_old_tool_outputs",
        "context_budget",
    )?;
    let prune_keep_recent_user_turns = expect_int(
        merged
            .get("prune_keep_recent_user_turns")
            .unwrap_or(&json!(DEFAULT_PRUNE_KEEP_RECENT_USER_TURNS)),
        "prune_keep_recent_user_turns",
        1,
        "context_budget",
    )?;
    let prune_protect_input_tokens = expect_int(
        merged
            .get("prune_protect_input_tokens")
            .unwrap_or(&json!(DEFAULT_PRUNE_PROTECT_INPUT_TOKENS)),
        "prune_protect_input_tokens",
        0,
        "context_budget",
    )?;
    let prune_min_input_tokens = expect_int(
        merged
            .get("prune_min_input_tokens")
            .unwrap_or(&json!(DEFAULT_PRUNE_MIN_INPUT_TOKENS)),
        "prune_min_input_tokens",
        0,
        "context_budget",
    )?;
    let compact_summary_max_output_tokens = expect_int(
        merged
            .get("compact_summary_max_output_tokens")
            .unwrap_or(&json!(DEFAULT_COMPACT_SUMMARY_MAX_OUTPUT_TOKENS)),
        "compact_summary_max_output_tokens",
        1,
        "context_budget",
    )?;
    let compact_refresh_min_new_messages = expect_int(
        merged
            .get("compact_refresh_min_new_messages")
            .unwrap_or(&json!(DEFAULT_COMPACT_REFRESH_MIN_NEW_MESSAGES)),
        "compact_refresh_min_new_messages",
        1,
        "context_budget",
    )?;
    let overflow_keep_recent_user_turns = expect_int(
        merged
            .get("overflow_keep_recent_user_turns")
            .unwrap_or(&json!(DEFAULT_OVERFLOW_KEEP_RECENT_USER_TURNS)),
        "overflow_keep_recent_user_turns",
        1,
        "context_budget",
    )?;
    let overflow_disable_tools_on_final_attempt = expect_bool(
        merged
            .get("overflow_disable_tools_on_final_attempt")
            .unwrap_or(&Value::Bool(true)),
        "overflow_disable_tools_on_final_attempt",
        "context_budget",
    )?;
    let overflow_final_max_output_default =
        model.map(|item| item.max_output.min(512)).unwrap_or(512);
    let overflow_final_max_output_tokens = expect_int(
        merged
            .get("overflow_final_max_output_tokens")
            .unwrap_or(&json!(overflow_final_max_output_default)),
        "overflow_final_max_output_tokens",
        1,
        "context_budget",
    )?;

    Ok(ContextBudgetOptions {
        enabled,
        strategy,
        counting,
        compaction_mode,
        reserve_output_tokens,
        guard_ratio,
        input_safety_margin_tokens,
        use_safety_margin_tokens,
        explicit_input_safety_margin_tokens,
        bytes_per_token,
        tool_display_max_bytes,
        tool_context_preview_bytes,
        tool_context_preview_lines,
        tool_context_line_max_chars,
        prune_old_tool_outputs,
        prune_keep_recent_user_turns,
        prune_protect_input_tokens,
        prune_min_input_tokens,
        compact_summary_max_output_tokens,
        compact_refresh_min_new_messages,
        overflow_keep_recent_user_turns,
        overflow_disable_tools_on_final_attempt,
        overflow_final_max_output_tokens,
    })
}

pub fn check_context_budget(
    system: Option<&str>,
    messages: &[ChatMessage],
    tools: &[ToolSchema],
    model: Option<&Model>,
    options: Option<&Value>,
    fallback_stage: &str,
) -> Result<Option<ContextBudgetResult>, String> {
    let Some(model) = model else {
        return Ok(None);
    };
    if model.context_window == 0 {
        return Ok(None);
    }
    let config = load_context_budget_options(options, Some(model))?;
    if !config.enabled {
        return Ok(None);
    }
    if !SUPPORTED_STRATEGIES.contains(&config.strategy.as_str()) {
        return Err(format!(
            "Unsupported context budget strategy: {}. Supported strategies: auto, error, compact.",
            config.strategy
        ));
    }
    if !SUPPORTED_COUNTING.contains(&config.counting.as_str()) {
        return Err(format!(
            "Unsupported context budget counting mode: {}. Supported modes: auto, tiktoken, heuristic.",
            config.counting
        ));
    }

    let provider_options = options_to_btree(options);
    let payload = materialize_openai_compatible_payload(
        system,
        messages,
        tools,
        Some(model),
        Some(&provider_options),
    );
    let payload_kind = if is_openai_compatible_model(model) {
        "openai_compatible"
    } else {
        "generic"
    };
    let count = estimate_payload_tokens(&payload, config.bytes_per_token);
    let diagnostics = tool_message_diagnostics(messages, model, options, config.bytes_per_token);
    let input_limit_tokens = compute_input_limit_tokens(model, &config);
    Ok(Some(ContextBudgetResult {
        estimated_input_tokens: count,
        input_limit_tokens,
        context_window: model.context_window,
        reserved_output_tokens: config.reserve_output_tokens,
        overflowed: count > input_limit_tokens,
        tool_message_count: diagnostics.tool_message_count,
        largest_tool_message_tokens: diagnostics.largest_tool_message_tokens,
        largest_tool_message_name: diagnostics.largest_tool_message_name,
        counting_method: "heuristic".to_string(),
        counting_exact: false,
        fallback_stage: fallback_stage.to_string(),
        payload_kind: payload_kind.to_string(),
    }))
}
